### PR TITLE
feat: dont search for spec files in `.gitignore` dirs

### PR DESCRIPTION
**Pull request by domharrington**
## 🧰 Changes

Right now if you run the following in a directory:

```sh
npm init -y
npm i @readme/micro@2.3.0
npx @readme/micro
```

You get the following output:

```sh
npx @readme/micro
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

SyntaxError: Supplied schema is not a valid OpenAPI definition.
    at OpenAPIParser.parse (/private/tmp/micro-action-test/node_modules/@readme/openapi-parser/lib/index.js:79:19)
    at async OpenAPIParser.resolve (/private/tmp/micro-action-test/node_modules/@readme/json-schema-ref-parser/lib/index.js:189:5)
    at async OpenAPIParser.bundle (/private/tmp/micro-action-test/node_modules/@readme/json-schema-ref-parser/lib/index.js:231:5)
    at async main (/private/tmp/micro-action-test/node_modules/@readme/micro/index.js:85:29) {
  toJSON: [Function: toJSON],
  [Symbol(nodejs.util.inspect.custom)]: [Function: inspect]
}
```

This is because we're globbing across the whole directory (including node_modules) and we're actually picking up the JSONSchema file for the OpenAPI spec itself located at `node_modules/@apidevtools/openapi-schemas/schemas/v3.0/schema.yaml`

This change brings in the [`ignore`](https://www.npmjs.com/package/ignore) package and checks to see if there's a gitignore in the current directory, and if there is then it adds it to the list of files that get excluded post-globbing.

## 🧬 QA & Testing

Do the tests pass? Can you run the following?

```sh
npm init -y
npm i "https://github.com/readmeio/readme-micro#feat/dont-search-gitignore-dirs"
echo "node_modules" > .gitignore
npx @readme/micro
```

It will now output:

```sh
$ npx @readme/micro
TypeError [ERR_HTTP_INVALID_HEADER_VALUE]: Invalid value "undefined" for header "X-API-KEY"
    at ClientRequest.setHeader (node:_http_outgoing:647:3)
    at new ClientRequest (node:_http_client:284:14)
    at Object.request (node:https:357:10)
    at RedirectableRequest._performRequest (/private/tmp/micro-action-test-2/node_modules/follow-redirects/index.js:284:24)
    at new RedirectableRequest (/private/tmp/micro-action-test-2/node_modules/follow-redirects/index.js:66:8)
    at Object.request (/private/tmp/micro-action-test-2/node_modules/follow-redirects/index.js:523:14)
    at dispatchHttpRequest (/private/tmp/micro-action-test-2/node_modules/axios/lib/adapters/http.js:262:25)
    at new Promise (<anonymous>)
    at httpAdapter (/private/tmp/micro-action-test-2/node_modules/axios/lib/adapters/http.js:49:10)
    at dispatchRequest (/private/tmp/micro-action-test-2/node_modules/axios/lib/core/dispatchRequest.js:58:10) {
  code: 'ERR_HTTP_INVALID_HEADER_VALUE'
}
::error::Invalid value "undefined" for header "X-API-KEY"
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

TypeError [ERR_HTTP_INVALID_HEADER_VALUE]: Invalid value "undefined" for header "X-API-KEY"
    at ClientRequest.setHeader (node:_http_outgoing:647:3)
    at new ClientRequest (node:_http_client:284:14)
    at Object.request (node:https:357:10)
    at RedirectableRequest._performRequest (/private/tmp/micro-action-test-2/node_modules/follow-redirects/index.js:284:24)
    at new RedirectableRequest (/private/tmp/micro-action-test-2/node_modules/follow-redirects/index.js:66:8)
    at Object.request (/private/tmp/micro-action-test-2/node_modules/follow-redirects/index.js:523:14)
    at dispatchHttpRequest (/private/tmp/micro-action-test-2/node_modules/axios/lib/adapters/http.js:262:25)
    at new Promise (<anonymous>)
    at httpAdapter (/private/tmp/micro-action-test-2/node_modules/axios/lib/adapters/http.js:49:10)
    at dispatchRequest (/private/tmp/micro-action-test-2/node_modules/axios/lib/core/dispatchRequest.js:58:10) {
  code: 'ERR_HTTP_INVALID_HEADER_VALUE'
}
```

Which is another issue!

### CODE REVIEW

File: utils.js [hunk 1]
```diff
@@ -46,13 +44,13 @@ module.exports = {
...
-      out = out.concat(list.filter(item => out.indexOf(item) < 0));
-    });
-    return out;
+    const GITIGNORE_LOCATION = path.join(process.cwd(), '.gitignore');
+    const ig = ignore();
+    if (fs.existsSync(GITIGNORE_LOCATION)) {
+      ig.add(fs.readFileSync(GITIGNORE_LOCATION).toString());
+    }
+
+    return glob.sync(globs).filter(ig.createFilter()).filter(filterOas);
```
Comment (cvan) [inline utils.js:53]: would using npm's pkg for gitignore/npmignore be easier to maintain?
https://github.com/npm/ignore-walk#ignore-walk

this works too - fixes the bug & unblocks

File: utils.js [hunk 2]
```diff
@@ -2,6 +2,7 @@ const fs = require('fs');
 const path = require('path');
 
 const glob = require('glob');
+const ignore = require('ignore');
```
Comment (cvan) [inline utils.js:5]: npm has a pkg specialised for gitignore/npmignore be easier to maintain?

https://github.com/npm/ignore-walk#ignore-walk

if it's easier to keep as is, no need

**Review by cvan — APPROVED** [review]
good catch & fix!

File: utils.js (hunk 2 above)
Comment (domharrington) [reply utils.js:5]: Nice idea, but that returns with a list of all files, which we'd then have to parse the user-provided glob and filter that list down again.

**Pull request merged**